### PR TITLE
Distributed Training of the Decoder

### PR DESCRIPTION
**Pull request by Veldrovive**
The same methods used in #112 are used here as well. Implementation is via [huggingface accelerate](https://github.com/huggingface/accelerate) to minimize changes needed to have a script capable of single or multi GPU training.

The method used was to convert the decoder trainer class to wrap its models with an accelerator object passed in as an argument. Functionality remains mostly unchanged. One new change was made to separate the loss values for the different unets during training and validation in order to make it easier to see which need more training.

This is not ready due to some remaining issues, but I wanted a place to track them and have others provide input.
Known Issues:
- [x] Evalaution metrics hang on distributed mode
- [x] Training loss drops slowly and arbitrarily diverges (This appears to also be an issue on the main branch)
- [x] Loading the model causes the optimizer to error

Roadmap:
- [x] Implement distributed `DecoderTrainer`
- [x] Update `train_decoder.py` to use the distributed trainer.
- [x] Fix remaining issues
- [x] Verify functionality on large training set
- [ ] Port script for imagen

### CODE REVIEW

File: dalle2_pytorch/dalle2_pytorch.py [hunk 1]
```diff
@@ -2057,7 +2058,7 @@ def sample(
 
         for unet_number, unet, vae, channel, image_size, predict_x_start, learned_variance in tqdm(zip(range(1, len(self.unets) + 1), self.unets, self.vaes, self.sample_channels, self.image_sizes, self.predict_x_start, self.learned_variance)):
 
-            context = self.one_unet_in_gpu(unet = unet) if is_cuda else null_context()
+            context = self.one_unet_in_gpu(unet = unet) if is_cuda and not distributed else null_context()
```
Comment (Veldrovive) [inline dalle2_pytorch/dalle2_pytorch.py:2061]: One downside of using accelerate instead of handling ddp at a low level is that accelerate does not like when submodels are modified in any way. This becomes an issue in the sample method when a unet is moved to a device. In order to get around this, we create a parameter that prevents this move if we are doing distributed training.

**Comment by nousr** [comment]
> Evalaution metrics hang on distributed mode

If you haven't already figured this one out, you might need to unwrap the model before doing evaluation. Not doing so can cause a deadlock.

> Loading the model causes the optimizer to error

Could be an issue with param groups, I had a lot of issues with loading the optimizer. 1ed0f9d80b097594e6294bad432d06ce3fdcb220 and if you can, try using `group_wd_params=False`, it seems to work for the prior, at least. (Note: we can try using an `OrderedDict` for grouping the `wd_params`, I haven't looked into it, but it may be more deterministic than the standard dictionary.

**Comment by Veldrovive** [comment]
The hanging on evaluation problem was caused by torchmetrics assuming evaluation is happening on all processes while I was only doing it on master. Using torchmetrics built in distributed mode causes errors with further dataloading, however. So one rabbit hole later I found that it was a problem with how torchmetrics does its gather and substituting my own gather using accelerate fixes the issue.

**Comment by Veldrovive** [comment]
> Could be an issue with param groups, I had a lot of issues with loading the optimizer. [1ed0f9d](https://github.com/lucidrains/DALLE2-pytorch/commit/1ed0f9d80b097594e6294bad432d06ce3fdcb220) and if you can, try using `group_wd_params=False`, it seems to work for the prior, at least. (Note: we can try using an `OrderedDict` for grouping the `wd_params`, I haven't looked into it, but it may be more deterministic than the standard dictionary.

@nousr Yea, that seems to have fixed the issue. I don't have a good handle on why the split results in non-deterministic behavior though. Is it just the conversion between sets and lists that is messing up the parameter order?

**Comment by nousr** [comment]
>  Is it just the conversion between sets and lists that is messing up the parameter order?

@Veldrovive oh shoot! yeah....good catch--I didn't even see those `set()` calls up there in `separate_weight_decayable_params()`. I was so hyperfocused on `get_optimizer()` that once it worked with `group_wd_params=False` I was happy enough to continue on with my life😆 

I bet changing those to `list()` would fix the issue, should be an easy test & PR (assuming it works) if you get some time.

**Comment by Veldrovive** [comment]
The optimizer has been fixed and tested in #130 so it should no longer be necessary to use the `group_wd_params=False` option.

**Comment by Veldrovive** [comment]
I brought the branch up with main and started a longer term test run with only the initial unet on 3 GPUs. You can track its progress [here](https://wandb.ai/veldrovive/dalle2_train_decoder/runs/1qywwsxb?workspace=user-veldrovive).

**Comment by Veldrovive** [comment]
The training run above converged about as quickly as we would expect given its distributed batch size and all evaluation, saving, and logging works.

**Comment by Veldrovive** [comment]
The change to model saving could be part of its own pull request for improving training features as it does not really fit under the umbrella of distributed training. I also want to change config logging to save the config dict as a file instead of in the wandb config object as that is more general so I could open a pull request for those features if we want to be good about keeping separate feature changes in separate requests. Up to whoever's merging if I should separate it out. I just needed the feature for testing in this branch.

I can also go and squash these commits if you do not do a squash during merge.

**Comment by lucidrains** [comment]
> The training run above converged about as quickly as we would expect given its distributed batch size and all evaluation, saving, and logging works.

with accelerate, would we still have the option to go single GPU if need be? (given the big changes to the trainer class)

**Comment by Veldrovive** [comment]
> with accelerate, would we still have the option to go single GPU if need be? (given the big changes to the trainer class)

That's actually my favorite feature of accelerate. If you launch with `python train_decoder.py` you get bog standard training on the first available device and accelerate basically does nothing. You do have to abide by the rules of accelerate, however, which really only means it is harder to select which device to train on since you have to use the `CUDA_VISIBLE_DEVICES` environment variable.

The simplicity does come with the cost that in order to use extra features like mixed precision you have to use `accelerate launch` since accelerate handles scaling itself, but it works so well in every situation that it feels worth the slight inconvenience.

**Comment by lucidrains** [comment]
very cool and great job! i'll review it this weekend and we can try to have it merged on Monday :)

**Comment by 1073521013** [comment]
> The training run above converged about as quickly as we would expect given its distributed batch size and all evaluation, saving, and logging works.

First of all, thanks very much about the contribution of your distributed training code. When running this decoder distributed training,   it will hang up in the line of **self.accelerator.backward(loss)** if make **find_unused_parameters=True**, it will hang up in the line of **loss_map = {f"Unet {index} Training Loss": loss.item() for index, loss in enumerate(unet_average_loss)
 if loss != 0}**  (The output result of **accelerator.gather(unet_losses_tensor)** cannot be iterated) if make find_unused_parameters=False. Thanks again, looking forward to your reply.

**Comment by Veldrovive** [comment]
> First of all, thanks very much about the contribution of your distributed training code. When running this decoder distributed training, it will hang up in the line of **self.accelerator.backward(loss)** if make **find_unused_parameters=True**, it will hang up in the line of **loss_map = {f"Unet {index} Training Loss": loss.item() for index, loss in enumerate(unet_average_loss) if loss != 0}** (The output result of **accelerator.gather(unet_losses_tensor)** cannot be iterated) if make find_unused_parameters=False. Thanks again, looking forward to your reply.

That is very strange behavior. I assume you are editing the line `ddp_kwargs = DistributedDataParallelKwargs(find_unused_parameters=True)`?

Setting `find_unused_parameters` to False should error on the backward pass because `unets.0.null_text_embed` does not participate in the loss calculation since text encodings are not supported. Did you modify the code to use text encodings?

I don't know of anything that would cause it to hang on the dict comprehension. If it is hanging, it should be on a gather call. Can you print out the intermediate value between the gather and comprehension?

**Comment by Veldrovive** [comment]
@1073521013 Ok, I don't have any idea what could be causing the issue. There are three options I see. It could be something I didn't account for that you are using in your config, a code change that doesn't seem like it should affect anything but still causes issues, or a torch/accelerate version mismatch. If you could post your config in a pastebin we can start eliminating options here.

**Comment by 1073521013** [comment]
> @1073521013 Ok, I don't have any idea what could be causing the issue. There are three options I see. It could be something I didn't account for that you are using in your config, a code change that doesn't seem like it should affect anything but still causes issues, or a torch/accelerate version mismatch. If you could post your config in a pastebin we can start eliminating options there.

torch/accelerate version has matched(same problem). and my config:
{
    "decoder": {
        "unets": [
            {
                "dim": 512,
                "image_embed_dim": 512,
                "cond_dim": null,
                "channels": 3,
                "dim_mults": [1, 2, 3, 4],
                "attn_dim_head": 64,
                "attn_heads": 32
            }
        ],
        "image_sizes": [64],
        "channels": 3,
        "timesteps": 1000,
        "loss_type": "l2",
        "beta_schedule": "cosine",
        "learned_variance": true
    },
    "data": {
        "webdataset_base_url": "/home/notebook/data/group/ma/laion400M_data/laion400m-met-release/laion400m-webdataset/{}.tar",
        "embeddings_url": "/home/notebook/data/group/laion_400_embeddings/laion400m-embeddings/img_emb",
        "num_workers": 2,
        "batch_size": 25,
        "start_shard": 0,
        "end_shard": 41512,
        "shard_width": 5,
        "index_width": 4,
        "splits": {
            "train": 0.9,
            "val": 0.05,
            "test": 0.05
        },
        "shuffle_train": true,
        "resample_train": false,
        "preprocessing": {
            "RandomResizedCrop": {
                "size": [64, 64],
                "scale": [0.75, 1.0],
                "ratio": [1.0, 1.0]
            },
            "ToTensor": true
        }
    },
    "train": {
        "epochs": 100,
        "lr": 2e-5,
        "wd": 0.1,
        "max_grad_norm": 0.5,
        "save_every_n_samples": 300000,
        "n_sample_images": 6,
        "device": "cuda:0",
        "epoch_samples": 2000000,
        "validation_samples": 50000,
        "use_ema": false,
        "ema_beta": 0.99,
        "amp": false,
        "save_all": false,
        "save_latest": true,
        "save_best": true,
        "unet_training_mask": [true]
    },
    "evaluate": {
        "n_evaluation_samples": 10,
        "FID": {
            "feature": 64
        },
        "IS": {
            "feature": 64,
            "splits": 10
        },
        "KID": {
            "feature": 64,
            "subset_size": 10
        },
        "LPIPS": {
            "net_type": "vgg",
            "reduction": "mean"
        }
    },
    "tracker": {
        "tracker_type": "wandb",
        "data_path": "./models_all",
        "image_path":"./wandb",
        "wandb_entity": "foreveryoung",
        "wandb_project": "decoder",
        "verbose": false
    },
    "load": {
        "source": null,
        "run_path": "foreveryoung/decoder/azotfo37",
        "file_path": "latest.pth",
        "resume": false
    }
}

**Comment by Veldrovive** [comment]
Ah, ok the most likely explanation is that 400M has not been reordered so the image shards do not line up with the embeddings shards. If that happens one process will load a shard correctly and another won't which leads to a hang on the first gather it gets to. The reason it doesn't error is because webdataset hangs instead of erroring if an image shard it's trying to load doesn't exist.  This is also a problem if you use an end shard that is greater than the number of shards. This is a good thing to bring up because it will be one of the most common issues and it is worth addressing with better logging.

Logging won't solve the base issue of the dataset not working for training the decoder, though. Which source are you using to get 400M? I can probably reorder the embeddings so that it is a valid dataset.

We should also add an option to do on the fly embedding generation so that if a dataset does not have reordered embeddings and you don't have the ability to generate them yourself you can trade off speed and just load images.

**Comment by 1073521013** [comment]
This is indeed the case. My data is obtained from [here ](https://deploy.laion.ai/8f83b608504d46bb81708ec86e912220/). Due to some other problems, I downloaded the training data locally for training. I originally wanted to code to align it, but I saw that you have implemented embedding-dataset-reordering, so I used your script to align it. After the final alignment, I experimented with it. There are still two small problems, one is after alignment The embedding npy file has a large number of 0 matrices (which may represent images that were not successfully downloaded), resulting in lots of problems like `UserWarning: RuntimeError('Webdataset had a sample, but no embedding was found. ImgShard: 10093 - Index: 0229')`; Another problem is that although the distributed operation is running, a step only executes one batch instead of gpus*batch. It is still under research, thanks.

**Comment by Veldrovive** [comment]
0 vectors are inserted into the embeddings whenever an image index was skipped when computing embeddings. That warning showing up in large quantities means a lot of image links were dead when embeddings were computed.

> Another problem is that although the distributed operation is running, a step only executes one batch instead of gpus*batch.

I am not sure exactly what this means. Are you saying that `sample` only increases by batch size every step instead of by gpus*batch size?

**Comment by 1073521013** [comment]
> This is indeed the case. My data is obtained from [here ](https://deploy.laion.ai/8f83b608504d46bb81708ec86e912220/). Due to some other problems, I downloaded the training data locally for training. I originally wanted to code to align it, but I saw that you have implemented embedding-dataset-reordering, so I used your script to align it. After the final alignment, I experimented with it. There are still two small problems, one is after alignment The embedding npy file has a large number of 0 matrices (which may represent images that were not successfully downloaded), resulting in lots of problems like `UserWarning: RuntimeError('Webdataset had a sample, but no embedding was found. ImgShard: 10093 - Index: 0229')`; Another problem is that although the distributed operation is running, a step only executes one batch instead of gpus*batch. It is still under research, thanks.

I carefully studied the data downloaded [before](https://deploy.laion.ai/8f83b608504d46bb81708ec86e912220/embeddings/), and found that the meta parquet data of `embeddings/metadata/` is different from the parquet log data in webdataset that the download is successful, and they are not aligned, which leads to a large number of warnings;
The second problem is that my code is wrong. and it's solved; Many thanks for your hospitality.

**Comment by YUHANG-Ma** [comment]
Hi, thanks for your contribution. When I use your code, it casues an error shown in the following pic.
50af5ab87cc689a27d4a8ceceb2d88a
do you know how to fix it?

**Comment by nousr** [comment]
> Hi, thanks for your contribution. When I use your code, it casues an error shown in the following pic.
> 
> 50af5ab87cc689a27d4a8ceceb2d88a
> 
> do you know how to fix it?

Hey! This issue happens when you try to call the member function of a Distributed model.

The common solution for this is "unwrapping" the model by calling `distributed_model.module.member_function()`. 

Hopefully this puts you on the right track!

**Comment by YUHANG-Ma** [comment]
> 

Hi, I am also confused. Where do I need to unwrapping the model?

**Comment by YUHANG-Ma** [comment]
> > Hi, thanks for your contribution. When I use your code, it casues an error shown in the following pic.
> > 50af5ab87cc689a27d4a8ceceb2d88a
> > do you know how to fix it?
> 
> Hey! This issue happens when you try to call the member function of a Distributed model.
> 
> The common solution for this is "unwrapping" the model by calling `distributed_model.module.member_function()`.
> 
> Hopefully this puts you on the right track!

image
Do you mean I need to fix it like this?

**Comment by YUHANG-Ma** [comment]
> > > Hi, thanks for your contribution. When I use your code, it casues an error shown in the following pic.
> > > 50af5ab87cc689a27d4a8ceceb2d88a
> > > do you know how to fix it?
> > 
> > 
> > Hey! This issue happens when you try to call the member function of a Distributed model.
> > The common solution for this is "unwrapping" the model by calling `distributed_model.module.member_function()`.
> > Hopefully this puts you on the right track!
> 
> image Do you mean I need to fix it like this?

Thanks, I fix it!

**Comment by YUHANG-Ma** [comment]
Hi, I also have an issue. My dataset contains about 17million data. But during my training process, in epoch 0 it samples more than 17million. Why did it happen?
image

**Comment by YUHANG-Ma** [comment]
> > Hi, thanks for your contribution. When I use your code, it casues an error shown in the following pic. 50af5ab87cc689a27d4a8ceceb2d88a do you know how to fix it?
> 
> That's on me. It is fixed in a commit I have not not pushed.

I also met a problem that yy dataset contains about 17million data, but during my training process, in epoch 0 it samples more than 17million. It couldn't move to epoch 1 and it failed t o save the checkpoint although I set the "save_all" as true in config.

**Comment by nousr** [comment]
> Hi, sorry to bother you again. I met an issue after I loaded the checkpoints. I can load the model successfully. But when it starts training, it shows an error. I am confused about it. Why did this happen even if the model is loaded successfully?
> 
> image
> 
> 

Hey! This looks like a problem with the optimizer. I would ensure that you are utilizing the same hyper-parameters for the decoder when you saved the checkpoint. Additionally, I would make sure you have the latest version of the repo.

If you've tried both of those and still get the error, please open an issue with a minimal example to recreate the problem so we can try to get to the bottom of it!

**Comment by YUHANG-Ma** [comment]
> > Hi, sorry to bother you again. I met an issue after I loaded the checkpoints. I can load the model successfully. But when it starts training, it shows an error. I am confused about it. Why did this happen even if the model is loaded successfully?
> > image
> 
> Hey! This looks like a problem with the optimizer. I would ensure that you are utilizing the same hyper-parameters for the decoder as when you saved the checkpoint. Additionally, I would make sure you have the latest version of the repo (and it might be worth checking that this issue still presents itself if you save a checkpoint with the latest version as well).
> 
> If you've tried both of those and still get the error, please open an issue with a minimal example to recreate the problem so we can try to get to the bottom of it!

Thanks, I will try to download the latest version of the repo. Is that from https://github.com/Veldrovive/DALLE2-pytorch/tree/decoder_trainer_accelerate?

**Comment by lucidrains** [comment]
@Veldrovive hey Aidan! thanks for continuing to push on this pull request!

we can merge this any time you think it is ready

(don't worry about Imagen btw, i will probably complete it next week solo)

**Comment by Veldrovive** [comment]
@lucidrains I brought it even with main to resolve merge conflicts yesterday so that it could be merged. Nousr and I just finished tracking down a bug that I thought might have been introduced, but it turned out it was just wandb having strange behavior. [This](https://wandb.ai/veldrovive/dalle2_train_decoder/runs/3d5rytsa) is a run using the up to date version of the branch with sparse attention enabled for the unet. It is ready to be merged. Would you like me to squash the commits or jut leave it as is?

**Comment by lucidrains** [comment]
@Veldrovive it is perfect! let's merge it :pray:

@nousr @Veldrovive what is both of your emails btw? could you send me an email at lucidrains@gmail.com so i can keep track?

File: train_decoder.py [hunk 2]
```diff
@@ -237,157 +255,216 @@ def train(
...
+            accelerator.print(print_ribbon(f"Starting Validation {epoch}", repeat=40))
+            last_val_sample = val_sample
+            val_sample_length_tensor = torch.zeros(1, dtype=torch.int, device=inference_device)
+            average_val_loss_tensor = torch.zeros(1, trainer.num_unets, dtype=torch.float, device=inference_device)
             timer = Timer()
-            for i, (img, emb, *_) in enumerate(dataloaders["val"]):
-                sample += img.shape[0]
+            accelerator.wait_for_everyone()
+            i = 0
+            for i, (img, emb, txt) in enumerate(dataloaders["val"]):
```
Comment (mrsndmn) [inline train_decoder.py:379]: Shouldn't we use here prepared dataloaders from `accelerate.prepare`?
